src/Microsoft.Health.Fhir.Api/Features/Context/FhirAuthenticationExceptionHandlerMiddleware.cs

```diff
@@ -0,0 +1,46 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.AspNetCore.Http;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Microsoft.Health.Fhir.Api.Features.Context
+{
+    /// <summary>
+    /// Middlware that runs before authentication middleware and after FhirRequestContextBeforeAuthenticationMiddleware
+    /// </summary>
+    public class FhirAuthenticationExceptionHandlerMiddleware
+    {
+        private readonly RequestDelegate _next;
+
+        public FhirAuthenticationExceptionHandlerMiddleware(
+            RequestDelegate next)
+        {
+            EnsureArg.IsNotNull(next, nameof(next));
+            _next = next;
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            try
+            {
+                await _next(context);
+            }
+            catch (S2SAuthenticationException ex)
+            {
+                if (ex.InnerException is SecurityTokenInvalidAudienceException || ex.InnerException is SecurityTokenInvalidIssuerException)
+                {
+                    throw ex.InnerException;
```

Comment:
Sorry I'm iterating here, but even better:

```suggestion
                    ExceptionDispatchInfo.Capture(ex.InnerException).Throw();
```

This should allow a similar behavior to what Fernando mentioned and restore the original execution context and call stack, instead of looking like it stopped in this Middleware.